src/main/java/org/jabref/gui/mergeentries/BatchEntryMergeWithFetchedDataAction.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    @Override
+    public void execute() {
+        LibraryTab libraryTab = tabSupplier.get();
+        if (libraryTab == null) {
+            LOGGER.error("Cannot perform batch merge: no library is open.");
+            return;
+        }
+
+        List<BibEntry> libraryEntries = libraryTab.getDatabase().getEntries();
+        if (libraryEntries.isEmpty()) {
+            notificationService.notify(Localization.lang("empty library"));
+            return;
+        }
+
+        MergeContext context = new MergeContext(
+                libraryTab,
+                libraryEntries,
+                new MergingIdBasedFetcher(preferences.getImportFormatPreferences()),
+                notificationService
+        );
+
+        taskExecutor.execute(createBackgroundTask(context));
+    }
+
+    private static BackgroundTask<List<String>> createBackgroundTask(MergeContext context) {
+        BackgroundTask<List<String>> task = new BackgroundTask<>() {
+            @Override
+            public List<String> call() {
+                return processMergeEntries(context, this);
+            }
+        };
+
+        task.setTitle(Localization.lang("Fetching and merging entries"));
+        task.showToUser(true);
+
+        task.onSuccess(updatedEntries ->
+                handleSuccess(updatedEntries, context));
+
+        task.onFailure(ex ->
+                handleFailure(ex, context.notificationService()));
+
+        return task;
+    }
+
+    private static List<String> processMergeEntries(MergeContext context, BackgroundTask<?> task) {
+        int totalEntries = context.entries().size();
+
+        for (int i = 0; i < totalEntries && !task.isCancelled(); i++) {
+            BibEntry libraryEntry = context.entries().get(i);
+            updateProgress(i, totalEntries, task);
+            fetchAndMergeEntry(context, libraryEntry);
+        }
+
+        finalizeCompoundEdit(context);
+        return context.updatedEntries();
+    }
+
+    private static void fetchAndMergeEntry(MergeContext context, BibEntry libraryEntry) {
+        LOGGER.debug("Processing library entry: {}", libraryEntry);
+
+        context.fetcher().fetchEntry(libraryEntry)
+               .filter(MergingIdBasedFetcher.FetcherResult::hasChanges)
+               .ifPresent(result -> Platform.runLater(() -> {
+                   MergeEntriesHelper.mergeEntries(result.mergedEntry(), libraryEntry, context.compoundEdit());
+                   libraryEntry.getCitationKey().ifPresent(context.updatedEntries()::add);
+               }));
+    }
+
+    private static void finalizeCompoundEdit(MergeContext context) {
+        if (!context.updatedEntries().isEmpty()) {
+            Platform.runLater(() -> {
+                context.compoundEdit().end();
+                context.libraryTab().getUndoManager().addEdit(context.compoundEdit());
+            });
+        }
+    }
+
+    private static void updateProgress(int currentIndex, int totalEntries, BackgroundTask<?> task) {
```

Comment:
`static` method are a strong indicator for a bad design.

Please, simply create a new class inherting from `BackgroundTask`. See `org.jabref.gui.externalfiles.UnlinkedFilesCrawler` for an example.